activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -1,28 +1,49 @@
 module Enumerable
-  # Calculates a sum from the elements.
+  # Enumerable#sum was added in Ruby 2.4 but it only works with Numeric elements
+  # when we omit an identity.
   #
-  #  payments.sum { |p| p.price * p.tax_rate }
-  #  payments.sum(&:price)
-  #
-  # The latter is a shortcut for:
-  #
-  #  payments.inject(0) { |sum, p| sum + p.price }
-  #
-  # It can also calculate the sum without the use of a block.
-  #
-  #  [5, 15, 10].sum # => 30
-  #  ['foo', 'bar'].sum # => "foobar"
-  #  [[1, 2], [3, 1, 5]].sum # => [1, 2, 3, 1, 5]
-  #
-  # The default sum of an empty list is zero. You can override this default:
-  #
-  #  [].sum(Payment.new(0)) { |i| i.amount } # => Payment.new(0)
-  def sum(identity = nil, &block)
-    if block_given?
-      map(&block).sum(identity)
-    else
-      sum = identity ? inject(identity, :+) : inject(:+)
-      sum || identity || 0
+  # We tried shimming it to attempt the fast native method, rescue TypeError,
+  # and fall back to the compatible implementation, but that's much slower than
+  # just calling the compat method in the first place.
+  if Enumerable.instance_methods(false).include?(:sum) && !((?a..?b).sum rescue false)
+    # We can't use Refinements here because Refinements with Module which will be prepended
+    # doesn't work well https://bugs.ruby-lang.org/issues/13446
+    alias :orig_sum :sum
```

Comment:
If we run only the test, it will pass.

```
$ bundle exec ruby -I test test/core_ext/enumerable_test.rb
Run options: --seed 41825

# Running:

...........

Finished in 0.001698s, 6477.8854 runs/s, 70667.8405 assertions/s.

11 runs, 120 assertions, 0 failures, 0 errors, 0 skips
```

If we apply following patch, the test fill fail :fearful: 

```diff
diff --git a/activesupport/test/core_ext/enumerable_test.rb b/activesupport/test/core_ext/enumerable_test.rb
index 4f1ab993b8..2a786fc874 100644
--- a/activesupport/test/core_ext/enumerable_test.rb
+++ b/activesupport/test/core_ext/enumerable_test.rb
@@ -1,6 +1,7 @@
 require "abstract_unit"
 require "active_support/core_ext/array"
 require "active_support/core_ext/enumerable"
+require "active_support/core_ext/object/json"

 Payment = Struct.new(:price)
 ExpandedPayment = Struct.new(:dollars, :cents)
```

```
  1) Error:
EnumerableTests#test_sums:
NoMethodError: undefined method `orig_sum' for #<EnumerableTests::GenericEnumerable:0x00000002898f58>
    /app/activesupport/lib/active_support/core_ext/enumerable.rb:35:in `sum'
    test/core_ext/enumerable_test.rb:51:in `test_sums'


  2) Error:
EnumerableTests#test_empty_sums:
NoMethodError: undefined method `orig_sum' for #<EnumerableTests::GenericEnumerable:0x00000002893238 @values=[]>
    /app/activesupport/lib/active_support/core_ext/enumerable.rb:35:in `sum'
    test/core_ext/enumerable_test.rb:97:in `test_empty_sums'


  3) Error:
EnumerableTests#test_range_sums:
NoMethodError: undefined method `orig_sum' for 1..4:Range
    /app/activesupport/lib/active_support/core_ext/enumerable.rb:35:in `sum'
    /app/activesupport/lib/active_support/core_ext/enumerable.rb:127:in `sum'
    test/core_ext/enumerable_test.rb:112:in `test_range_sums'

11 runs, 83 assertions, 0 failures, 3 errors, 0 skips
```

The cause is using refinements with [prepend](https://github.com/rails/rails/blob/32a1c30f3893286ba048c3b72d5d4bcc8f520311/activesupport/lib/active_support/core_ext/object/json.rb#L45-L47).

```ruby
using Module.new {
  refine Enumerable do
    alias :orig_sum :sum
  end
}

module Enumerable
  def sum(*args)
    orig_sum(*args)
  end
end

class GenericEnumerable
  include Enumerable

  def each
  end
end

# GenericEnumerable.new.sum # if we uncomment this line, `GenericEnumerable#sum` will work
Enumerable.prepend(Module.new) # if we comment out this line, `GenericEnumerable#sum` will work

p GenericEnumerable.new.sum # undefined method `orig_sum' for #<GenericEnumerable:0x0000000127c120 @values=[1, 2, 3]> (NoMethodError)
```

It may be not intended behavior (I created the tickect https://bugs.ruby-lang.org/issues/13446).

Anyway, we can't use refinements here :cry: